BE-04 Add API endpoints and backend enviroment files

Rix95 opened the pull request:
- **BE-04: Add FASTSAPI to Docker**
- **BE-04: Modify Docker compose to add connectivity between api and database**
- **BE-04: Add Neo4j driver and config file to simplify loading enviroment variables. Add some api code template**

## Describe your changes
Add two partially functional endpoints (Add Repository, get list of repositories)
## Is this a fix to an open issue, if so please share link?
N/A
## Has this changes been tested?
docker compose up --build
and Api calls.

e-kari committed:
BE-03: Set up script to create neo4j schema

e-kari committed:
BE-03: Re-organizing files

Rix95 committed:
BE05: Add neo4j container

Rix95 committed:
BE-04: Modify Docker compose to add connectivity between api and database

Rix95 committed:
BE-04: Modify Docker compose to add connectivity between api and database

Rix95 committed:
BE-04: Add Neo4j driver and config file to simplify loading enviroment variables. Add some api code template

Rix95 committed:
Merge branch 'main' into api

richiekrich commented on src/backend/api/main.py:
```diff
@@ -7,5 +24,107 @@ def main():
+def get_vulnerability(repository: str="all"): #default behavior should retrieve total number of all repositories in all repos.
+    return None
+
 if __name__ == "__main__":
```
### Application Start-up:
The `if __name__ == "__main__":` block calls `main()`, which is the route handler and not the server startup command. When running the API, the server should be started with a command like:

```sh
uvicorn main:app ...
```
This block might be redundant if you always use the Docker container or an external command to launch the app.

richiekrich commented on src/backend/api/vulnerability_repository.py:
```diff
@@ -0,0 +1,65 @@
+    try:
+        with Neo4jDriver.get_driver().session() as session:
+            query = """
+            CREATE (r:Repository {
```
## Issues

### Label Inconsistency

The `create_repository_in_neo4j` function uses the label `Repository`, while the update and existence functions use `VulnerabilityRepository`. This mismatch will cause inconsistencies in data retrieval and update operations.

**Recommendation:**  
Standardize on one label (likely `VulnerabilityRepository` to be consistent with the GET endpoint in `main.py`).

richiekrich commented on docker-compose.yml:
## Issues

### Volume Definition

- Under the FastAPI service (in line 15) there’s an extra volume entry (`- /app/.venv`), which appears mis-indented or misformatted.  
- At the bottom, the volumes for Neo4j are duplicated. This redundancy might cause confusion and should be cleaned up.

richiekrich commented on docker-compose.yml:
```diff
@@ -38,12 +36,22 @@ services:
         - neo4j_import:/import
         - neo4j_plugins:/plugins
+      healthcheck:
+        test: ["CMD", "cypher-shell", "-u", "neo4j", "-p", "yourpassword", "RETURN 1"]
```
### Healthcheck Password

The Neo4j service’s healthcheck uses a hardcoded password (`yourpassword`). Ensure this aligns with the credentials defined in the environment files and configuration.

richiekrich reviewed (changes requested):
## Overall Recommendations

### Fix Environment Variable Typo
- Update `NEO4J_USERNMAE` to `NEO4J_USERNAME` in `main.py`.

### Standardize API Route Paths
- Add leading slashes to the routes for `add_vulnerability`, `get_vulnerability`, and `get_total_vulnerabilities`.

### Consistent Labeling in Neo4j Queries
- Use a single label (preferably `VulnerabilityRepository`) across create, update, and query operations.

### Review Docker Compose Volumes
- Correct the mis-indented volume mapping in the FastAPI service and remove duplicate volume definitions.

### Lock File Visibility
- Keep `uv.lock` committed as it aids reproducibility and does not expose sensitive information.

### ✏️ Note: 
- Comments have been added to the PR files to indicate where changes are needed.

richiekrich committed:
BE-04: Update api url

Owner: Richard Rich

Code Changes:
- In docker-compose.yml updated the variable to `http://fastapi:8000` in frontend container to point to backend api for consistency

richiekrich committed:
Merge pull request #30 from Rix95/main

update api branch with main branch's changes

Rix95 committed:
"small commit SQUASH"

Rix95 committed:
"squash"

Rix95 committed:
BE-04: Refactor fASTAPI Modules

Rix95 committed:
BE-04: Refactor fASTAPI Modules

Rix95 committed:
commit

Rix95 committed:
Merge branch 'main' into api

Rix95 committed:
temporal changes squash commit

richiekrich committed:
Merge pull request #37 from Rix95/main

update branch with main's changes

Commit:
BE-06: Fix Docker configuration and FastAPI connectivity
